File: core/src/main/java/com/gmail/marcosav2010/cipher/CipherPool.java
package com.gmail.marcosav2010.cipher;

import javax.crypto.Cipher;
import javax.crypto.spec.OAEPParameterSpec;
import javax.crypto.spec.PSource;
import java.security.GeneralSecurityException;
import java.security.Key;
import java.security.spec.AlgorithmParameterSpec;
import java.security.spec.MGF1ParameterSpec;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.atomic.AtomicInteger;

public class CipherPool {

    private static final int DEFAULT_MAX_CAPACITY = 10;
    private static final int DEFAULT_INITIAL_CIPHERS = 2;

    private static final String CIPHER_ASYMMETRIC_ALGORITHM = "RSA/ECB/OAEPWithSHA-512AndMGF1Padding";
    private static final AlgorithmParameterSpec ALGORITHM_SPEC = new OAEPParameterSpec("SHA3-512", "MGF1",
            MGF1ParameterSpec.SHA512, PSource.PSpecified.DEFAULT);
    private final int mode;
    private final Key key;
    private final int maxCapacity;
    private final BlockingQueue<Cipher> queue;
    private final AtomicInteger cipherCount;

    public CipherPool(int mode, Key key) throws GeneralSecurityException {
        this(mode, key, DEFAULT_INITIAL_CIPHERS, DEFAULT_MAX_CAPACITY);
    }

    public CipherPool(int mode, Key key, int initialCiphers, int maxCapacity) throws GeneralSecurityException {
        this.mode = mode;
        this.key = key;

        this.maxCapacity = maxCapacity;

        cipherCount = new AtomicInteger();

        queue = new ArrayBlockingQueue<>(maxCapacity);

        init(initialCiphers);
    }

    private void init(int count) throws GeneralSecurityException {
        if (count > maxCapacity)
            throw new IllegalArgumentException("Initial cipher number can't be higher than max ciphers");

        for (int i = 0; i < count; i++)
            queue.add(createCipher());
    }

    private Cipher createCipher() throws GeneralSecurityException {
        cipherCount.updateAndGet(i -> {
            if (++i > maxCapacity)
                throw new IllegalStateException("Cipher pool exceded cipher of " + maxCapacity);
            return i;
        });

        Cipher c = Cipher.getInstance(CIPHER_ASYMMETRIC_ALGORITHM);
        c.init(mode, key, ALGORITHM_SPEC);
        return c;
    }

    public final byte[] doFinal(byte[] bytes) throws GeneralSecurityException, InterruptedException {
        Cipher c;
        if (!queue.isEmpty())
            c = queue.take();
        else if (cipherCount.get() < maxCapacity)
            try {
                c = createCipher();
            } catch (IllegalStateException ex) {
                c = queue.take();
            }
        else
            c = queue.take();

        byte[] out = c.doFinal(bytes);
        queue.put(c);
        return out;
    }
}

File: core/src/main/java/com/gmail/marcosav2010/cipher/CipherPool.kt
package com.gmail.marcosav2010.cipher

import java.security.Key
import java.security.spec.AlgorithmParameterSpec
import java.security.spec.MGF1ParameterSpec
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue
import java.util.concurrent.atomic.AtomicInteger
import javax.crypto.Cipher
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PSource

class CipherPool(private val mode: Int,
                 private val key: Key,
                 initialCiphers: Int = DEFAULT_INITIAL_CIPHERS,
                 private val maxCapacity: Int = DEFAULT_MAX_CAPACITY) {

    companion object {
        private const val DEFAULT_MAX_CAPACITY = 10
        private const val DEFAULT_INITIAL_CIPHERS = 2
        private const val CIPHER_ASYMMETRIC_ALGORITHM = "RSA/ECB/OAEPWithSHA-512AndMGF1Padding"
        private val ALGORITHM_SPEC: AlgorithmParameterSpec = OAEPParameterSpec("SHA-512", "MGF1",
                MGF1ParameterSpec.SHA512, PSource.PSpecified.DEFAULT)
    }

    private val queue: BlockingQueue<Cipher> = ArrayBlockingQueue(maxCapacity)
    private val cipherCount: AtomicInteger = AtomicInteger()

    init {
        init(initialCiphers)
    }

    private fun init(count: Int) {
        require(count <= maxCapacity) { "Initial cipher number can't be higher than max ciphers" }
        for (i in 0 until count) queue.add(createCipher())
    }

    private fun createCipher(): Cipher {
        cipherCount.updateAndGet { i ->
            check(i + 1 <= maxCapacity) { "Cipher pool exceeded cipher of $maxCapacity" }
            i + 1
        }
        val c = Cipher.getInstance(CIPHER_ASYMMETRIC_ALGORITHM)
        c.init(mode, key, ALGORITHM_SPEC)
        return c
    }

    fun doFinal(bytes: ByteArray): ByteArray {
        val c: Cipher = if (!queue.isEmpty()) queue.take()
        else
            if (cipherCount.get() < maxCapacity) try {
                createCipher()
            } catch (ex: IllegalStateException) {
                queue.take()
            } else queue.take()

        val out = c.doFinal(bytes)
        queue.put(c)
        return out
    }
}